File: alien_cpp/code.cpp
/*
drummer2322
corbin.mc96@gmail.com
GCJ <year> <round> <problem>
<date>
Status: can be optimized by swithing vector to map
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

int main () 
{
	int L, D, N;

	std::cin >> L;
	std::cin >> D;
	std::cin >> N;
	std::cin.ignore();

	std::string dictionary [D];
	for (int i=0; i<D; i++) {
		std::getline(std::cin, dictionary[i]);
	}

	for (int tt=1; tt<=N; tt++) {
		std::cerr << tt << std::endl;

		std::string case_word;
		std::getline(std::cin, case_word);

		int current_letter = 0;
		
		std::vector <int> invalid_indexes;

		while (case_word.length() > 0) {
			std::vector <char> required_letters;
			if (case_word[0] != '(') {
				required_letters.push_back(case_word[0]);
				case_word.erase(case_word.begin());
			}
			
			else {
				std::string::iterator closing_parenthesis = std::find(case_word.begin(), case_word.end(), ')');
				std::string::iterator p = case_word.begin()+1;
				while (p != closing_parenthesis) {
					required_letters.push_back(*(p++));
				}
				case_word.erase(case_word.begin(), closing_parenthesis+1);
			}

			for (int dictionary_word=0; dictionary_word<D; dictionary_word++) {
				if (std::find(invalid_indexes.begin(), invalid_indexes.end(), dictionary_word) != invalid_indexes.end()) {
					continue;				
				}
				if (std::find(required_letters.begin(), required_letters.end(), dictionary[dictionary_word][current_letter]) == required_letters.end()) {
					invalid_indexes.push_back(dictionary_word);
				}
			}

			current_letter++;
		}
		std::cout << "Case #" << tt << ": " << D - invalid_indexes.size() << std::endl;
	}
}

File: alien_py/code.py
# drummer2322
# corbin.mc96@gmail.com
# GCJ <year> <round> <problem>
# <date>

from sys import argv

if len(argv) > 1:
	type_of_case = argv[1]+'_'
else:
	type_of_case = ''
lines = [line[:-1] for line in open(type_of_case+'input.in')]
output = open(type_of_case+'output.out','w')

L, D, N = map(int, lines.pop(0).split());
permanent_dict = [];
for i in range(D):
	permanent_dict.append(lines.pop(0));

for t in range(1,N+1):
	print t

	case_dict = permanent_dict[:];	

	case_word = lines.pop(0);

	while not case_word=="":
		if case_word[0] != '(':
			required_letters=[case_word[0]];
			case_word=case_word[1:];
		else:
			closing_location = case_word.find(")");
			required_letters = case_word[1:closing_location];
			case_word=case_word[closing_location+1:]

		indexes_to_be_deleted = [];

		for i in range(len(case_dict)):
			if not case_dict[i][0] in required_letters:
				indexes_to_be_deleted.append(i);
			else:
				case_dict[i]=case_dict[i][1:];

		for i in indexes_to_be_deleted[::-1]:
			case_dict.pop(i);

	output.write("Case #%i: %s\n"%(t, len(case_dict) ))

output.close()
